### PR TITLE
Swap order of resource to be updated depended resource to be destroyed

### PR DESCRIPTION
**Pull request by magodo**
For **resA** which depends on **resB** on creation. Then if **resB** is deleted and **resA** is updated accordingly, the actual order user expected (as mentioned in issues mentioned in #23252) is: `update resA -> delete resB`.

This PR just swap the dependency order for the creator and destroyer in destroy edge transformer, and also modify the test accordingly.

### CODE REVIEW

**Comment by magodo** [comment]
@apparentlymart Hi Mart, is there any chance to get this PR reviewed?

**Comment by tustvold** [comment]
The current behavior is not only counter-intuitive, but it also creates issues for resources where the deletion cannot proceed until the update is performed. I've seen this specifically with datadog SLOs and monitors, but there are likely lots of other similar situations that would stand to benefit from this change :+1:

**Comment by magodo** [comment]
@tustvold Thank you for verifying that the issue still remains.
Hi @paultyng, any chance to take a look at this?

**Comment by pkolyvas** [comment]
Thanks for taking the time to submit this PR.

Unfortunately, we're going to have to turn it down. The behaviour of the PR is  incorrect for Terraform as it is today. Following the test file updated in the PR, the [very next test](https://github.com/hashicorp/terraform/blob/master/terraform/graph_builder_apply_test.go#L700) is for the lifecycle meta-argument `create_before_destroy` which defines the ordering you're trying to achieve. 

Resources which behave best when `create_before_destroy` is set can benefit from having such a suggestion added to their documentation.

**Comment by magodo** [comment]
@pkolyvas Thank you for your reply. I have two follow-up questions:
1. Is there any example that convinces me *The behaviour of the PR is incorrect for Terraform as it is today*?
2. As the expected order I want could be achieved by CBD, does that mean this is a side-effect of CBD? Plus there are [limitations](https://www.terraform.io/docs/configuration/resources.html#create_before_destroy) of CBD, it seems not a perfect solution to the **update from orphan** case.

**Comment by jbardin** [comment]
@magodo 

The simplest explanation is that destroying a dependency is only a reduction of the replacement scenario, and that the order of operations must be maintained between both.

For example, if `B` depends on `A`; then the replacement ordering is `A_destroy, A_create, B_update`, in which case you cannot have `B_update` depend on `A_destroy` without creating a simple cycle.  In order to be able to update `B` before `A` is destroyed, `A` must be created first, hence the need for `create_before_destroy` in this scenario. 

While this could be considered a side-effect of CBD, it is definitely part of the primary use case for `create_before_destroy`, and a reason why many resources document that their dependencies should be used with `create_before_destroy = true`.

**Comment by magodo** [comment]
@jbardin Thank you for providing the details above :+1: 
Actually the CBD doesn't work in my case, (following your example above) as when `A` is removed from terraform config, there is ed no way for terraform to know whether `A` is marked as CBD or not, so the order is still `A_destroy` -> `B_update`. This is slightly different than the replacement scenario.
Based on this point, i.e. the **destroying a dependency** is a special use case than **dependency replacement**, is it possible to specially handle that in the graph builder apply?

**Comment by jbardin** [comment]
We cannot change the order here, because there are also resources that need the ordering to be consistent. Keeping this consistent helps ensure predictability, regardless of the changes being made. 

The status of `create_before_destroy` is saved into the state in 0.13 specifically for this particular special use case, and the destroy order will be honored even when removing the resource from the config.

**Comment by magodo** [comment]
@jbardin Got it, many thanks! One minor reminder is to ensure mentioning this in the CBD documentation or the release note when v0.13 is released.

**Comment by magodo** [comment]
Hey @jbardin, after trying out in 0.13, it turns out for the case that `A` is an existing resource, whilst without CBD defined. Then user wants to remove the dependency of `A` from `B`, and remove `A` simultaneously, then user can't get the expected order since the CBD is not specified.
Currently, there seems to be no way for user to add the CBD on `A` at the moment that the user wants to remove it, except trigger some update on `A` so that CBD could be picked in state file.
Another way to address this issue is to provide some state subcommand to allow user to manipulate lifecycle metadata in statefile (similar to `terraform state mv`).
In summary, the only way to get the `A_destroy -> B_update` order is to ensure the CBD is specified at the time `A` is most recently updated/created. There lacks support for a one-step setting to get that order for `A` that has not `CBD` defined at the time to be destroyed.

**Comment by jbardin** [comment]
@magodo, thanks! I opened a new issue for you regarding the modification of existing resources.
